### ISSUE TITLE
man: document concept of dependencies and reverse dependencies in systemd(1)

### ISSUE THREAD

**Issue by thx1111** [systemd/systemd#720]
The documentation at man 5 systemd.unit says:

PartOf=
... When systemd stops or restarts the units listed here, the action is propagated to this unit. ...

However, developer "heftig" says, in Issue #615, 'Removing a device makes the unit disappear without actively "stopping" it. You want BindTo, not PartOf.'  This statement refers to undefined or misused terminology.  Superficially, the statement is completely incomprehensible, with respect to the phrase "makes the unit disappear".

A "Unit" is a kind of thing that is apparently defined implicitly, in "man 5 systemd.unit", in context with the repeated phrase "the units listed here", associated with the options "Requires=", "Requisite=", "RequisiteOverridable=", "Wants=", and especially, "PartOf=", and then, through example, referring to the "Unit Files" found in "/usr/lib/systemd/system/", where it can be discovered that the phrase "the units listed here" refers to a kind of _file_,  presumably, as also referenced immediately in "man 5 systemd.unit", under "DESCRIPTION", in the phrase "A unit configuration file encodes...".  Apparently, then, a "unit" is a kind of _file_.

But then, in the phrase "If this unit gets activated" under "Requires=", and in the phrase "stops or starts the units listed here" under "PartOf", these phrases clearly do not refer to a kind of "file", which is not a kind of thing that would be "activated", "started", or "stopped".

Referring back to the statement made by "heftig", 'Removing a device makes the unit disappear...', clearly, unit files do not "disappear",  are not "deleted", when a device, as for example a pccard, is removed.

Seemingly, then, a "unit" is a kind of thing that can be "activated", "started", "stopped", or be made to "disappear", but is also a kind of "file".  A reasonable person will infer that the term "unit" is being used to describe _TWO_ COMPLETELY DIFFERENT kinds of things.  This "double" use of terminology is NOT APPROPRIATE in documentation.

I have no doubt that the systemd developers understand the inner workings of systemd.  Yet the utility of systemd REQUIRES that THE REST OF US also understand how systemd is suppose to work.  To that end, it is necessary to provide comprehensible documentation, using definite terminology.

While I believe that the documentation conveys some idea of what a "Unit Configuration File" would be, specifically in the phrase "A unit configuration file encodes..." in "man 5 systemd.unit", I can find no specific definition for a kind of thing that can be "activated", "started", "stopped", and "disappeared".

WTF IS A "UNIT"???!!!

Let me take a guess:  A "Unit", not to be confused with a "Unit Configuration File" or identically, a "Unit File", IS AN INSTANCE OF 1) an executing computer program, or, 2) an inter-process communication mechanism of a type limited to file, socket, and named pipe type mechanisms.  File type IPC mechanisms includes both regular file system files, such as file paths and mount points, and special file system files, especially device files, such as partitions, and sysfs files, such as files for devices without device files and files for cgroups processes.

Is that correct?  And, if so, it should be in the documentation!

With respect to the "PartOf=" Unit File option, if I suppose that a sysfs file associated with a pccard "disappears" upon removal of that pccard - where the sysfs special file is literally deleted - _but_also_ does not "stop" the sysfs file associated with the pccard device, THEN THERE IS A MATERIAL DISTINCTION BETWEEN "disappearing a file" and "stopping a process".  It may be that the "PartOf=" option ONLY applies to processes, and NOT to IPC mechanisms.  Is that correct?  If so, that should be explicitly articulated in the documentation!

With respect to the "BindsTo=" Unit File option, the documentation says "it also declares that this unit is stopped when any of the units listed suddenly disappears."  Notice what the documentation does NOT say: "It also declares that this unit is STARTED when any of the units listed suddenly RE-APPEARS."

So then, HOW is this "reverse dependency" to be created?  It is NOT created as a "Unit Section" option.  Apparently, as a policy, systemd does NOT START a process with a "reverse dependency" UNLESS there exists an "Install Section" option "WantedBy=", "RequiredBy=", or "Alias=".  No process "starts" without being explicitly "enabled" by the system administrator.  This is good thing.  BUT - the policy is not discussed explicitly, and, for this "reverse dependency" to take effect, the Unit File has to be modified, reloaded, disabled,  and re-enabled.

Furthermore, the Unit Section option "Requires=" uses the terminology "activated" and "deactivated", "started" and "stopped".  It may be that the "Requires=" option, similarly with the "PartOf=" option, ONLY applies to processes, and NOT to IPC mechanisms.  If so, "Requires=" would be useless when referencing a ".device" file, and systemd WILL NOT provide any kind of error message or warning message referring to that usage.

THIS IS NOT DISCUSSED in the "BindsTo=" documentation.  Instead, the reader is left to deduce this "reverse dependency" configuration process after first "memorizing" and then "mentally processing" the entire "man 5 systemd.unit" page, while "guessing" about terminology.  This "reverse dependency" configuration process should be explicitly articulated in the "BindsTo=" documentation!

WHY does this "PartOf=" and "BindsTo=" documentation issue  matter?

Many, many users have wireless devices which rely upon wpa_supplicant, and implicitly, upon the systemd wpa_supplicant Service File.  Admittedly, not every user has a _removable_ wireless device.  Nonetheless, it should be expected that such a commonly used file, the wpa_supplicant Service File should be configured to function in a "robust" and generally effective fashion, rather than in just limited "optimal" circumstances.  It appears that, in order to "start" and "stop" the wpa_supplicant process with the "disappearance" and "re-appearance" of a wireless device, the Service File _must_ include:

[Unit]
BindsTo=sys-subsystem-net-devices-%i.device
After=sys-subsystem-net-devices-%i.device
...
[Install]
WantedBy=sys-subsystem-net-devices-%i.device

Is that correct?  At least I can say that, with those options, it seems to work as intended.

In contrast, the current "wpa_supplicant-2.4/wpa_supplicant/systemd/wpa_supplicant-nl80211.service.arg.in" file contains, in part:

[Unit]
Requires=sys-subsystem-net-devices-%i.device
After=sys-subsystem-net-devices-%i.device
...
[Install]
Alias=multi-user.target.wants/wpa_supplicant-nl80211@%i.service

where, then, the "Requires=" dependency is useless and does nothing, and only the "Alias=" dependency causes wpa_supplicant to even start.  The only actual process dependency is in "After=", and there is NOTHING otherwise connecting wpa_supplicant to the wireless device and nothing to start or stop wpa_supplicant when the actual wireless device is removable.

Is that correct?  That certainly seems to be the way it works.

In particular, wpa_supplicant responds very poorly to a wireless device that just "re-appears", and gratuitously starting wpa_supplicant after "multi-user.target" - and just "hoping it works" - is a poor plan.

It seems to me that these kinds of Unit File errors would be less common with corrections to the "PartOf=" and "BindsTo=" documentation.  This is important.

1) Explicitly define the term "Unit".
2) Distinguish between the term "Unit" making reference to 1) Processes and 2) IPC mechanisms.
3) Note which Unit File options apply _only_ to Processes and which apply _only_ to IPC mechanisms.
4) Explicitly mention the systemd policy, that a "reverse dependency" _starting_ a Process will _require_ an "Install Section" option, "WantedBy=", "RequiredBy=", or "Alias=".
5) Revise the descriptions for the "BindsTo=" and "PartOf=" options.

**Comment by poettering** [comment]
There's no need to curse, also, please try to stay concise and keep bug reports minimal, we all have enough to do, and reading page long rants is not what we want to spend our time on. thanks!

The basic concepts of systemd are explained on the main systemd man page systemd(1). There you'll find details about the "unit" concept. It currently lacks an explanation of the "reverse dependency" concept though.

The documentation of BindsTo= and PartOf= appears accurate and complete to me as it is.

I am not sure I follow with the rest of what you say, please read up on systemd(1) and the basic systemd concepts, and then make specific suggestions.

**Comment by thx1111** [comment]
Thanks for reviewing this issue Lennart.

> please read up on systemd(1) and the basic systemd concepts, and then make specific suggestions.

I've gone back and looked through systemd(1). The systemd(1) page delineates the set of meanings of the term "Unit" implicitly, but _not_ explicitly.  I still find this confusing.  The "concepts" section should be moved to systemd.unit(1), and then modified to include an _explicit_ definition of "Unit", as a preamble.

---

As used in systemd, the term "unit" is an "overloaded" term used in various contexts to refer to an instance of a class from only the following set of items, each representing the class:
1) A Unit Configuration File found in various directories and having a file name ending with .service, .socket, .device, .mount, .automount, .swap, .target, .path, .timer, .snapshot, .slice, or .scope.   "Unit" also refers to a type of section header in a Unit Configuration File.
2) An executing computer program, a Process.
3) An Inter-Process Communication Mechanism of a type limited to file, socket, and named pipe type mechanisms. File type IPC mechanisms includes both regular file system files such as file paths and mount points, and special file system files, especially device files such as partitions, and sysfs files such as files associated with devices that do not have device files, and sysfs files for cgroups processes.
4) A systemd observed State, representing the execution state of a process, the state of an IPC mechanism including the existence of a socket or device or the mount state of a file system, and the state of a systemd managed timer, snapshot, or control group.

The naming of Units in their different contexts is described in the man page associated with each type of Unit.

Generally, systemd will manage the state of Units, and Unit Configuration Files will describe "dependencies" between these systemd managed "activation states" - but with caveats.  The nature of a "activation state" depends upon the nature of the unit.  Processes and Timers can be "started" and "stopped".  File systems can be "mounted" and "unmounted".  Snapshots can be "saved" and "restored".  Devices can "appear" and "disappear".

"Activation states" can be made to propagate from one unit to another in "forward" and "reverse" directions.  A Unit Configuration File, then, can describe eight kinds of "dependencies"  between these "activation states", listed here with "A" the "Subject" Unit of the Unit Configuration File and "B" some other "Object" Unit, and where "Up" and "Down" represent a generalized concept of state:

Predicate : Consequence
Forward Dependencies -
A Up : B Up
A Up : B Down
A Down : B Up
A Down : B Down
Reverse Dependencies -
B Up : A Up
B Up : A Down
B Down : A Up
B Down : A Down

Of course, then, there exist two ways to configure dependencies, 1) as forward and reverse dependencies in the Subject Unit Configuration File, and 2) as forward and reverse dependencies in the Object Unit Configuration File.  Note that some "Units" managed by systemd, such as device interfaces, do not have explicit Unit Configuration Files, leaving only one file in which to define the dependency.  In other cases, the administrator may want to define dependencies in multiple Unit Configuration Files, in which case then systemd will disclose any inconsistency.

The options described below represent various combinations of these dependencies, with additional nuances, distinguishing the source or cause of a state transition.  For instance, a unit that is "stopped" by systemd is distinguished from a unit that "disappears" for some other reason such as a process that crashes or a device that is physically unplugged.

---

> The documentation of BindsTo= and PartOf= appears accurate and complete to me as it is.

Of course it does - you _wrote_ systemd. ;) But you also make automatic assumptions that other people do not make, and that the _reader_ will not make.  Your are too familiar with systemd to see it the way the "unitiated" will see it.  In particular, _without_ Jan's added comment, 'Removing a device makes the unit disappear without actively "stopping" it.', the current documentation for "PartOf=" is completely useless, and technically wrong, in that it is incomplete in a material way.  And again, Jan's use of the term "Unit" in this context, in reference to a sysfs filename, will make perfect sense to you and Jan, and make no sense at all to someone who is imagining that a "Unit" is a kind of executing computer program, a Unix "process", or that the term "Unit" is referring to a "Unit Configuration File", which would be the obvious default inference in this context, where the documentation literally says "the units listed here", meaning, literally, a "Unit Configuration File" in the list.  "Overloading" terminology is very expedient as lingo, and also very counterproductive when used in documentation.  The "PartOf=" description should be modified to include Jan's comment _and_ describe the need for an Install section "WantedBy=" dependency, in order to get the otherwise expected result with a "device" type "Unit".

---

PartOf=
Configures dependencies similar to Requires=, but limited to stopping and restarting of units. When systemd stops or restarts the units listed here, the action is propagated to this unit. Note that this is a one-way dependency — changes to this unit do not affect the listed units.  Also note that "removing" a device makes the corresponding sysfs file for that device "disappear" without the device actually having been "stopped" by systemd.  This option is not useful for a "device" type unit.  To "start" and "stop" this unit in sync with a "device" type unit, instead, use "BindsTo=" to "stop" this unit, and use "After=" in the Unit Section and "WantedBy=" in the Install Section to "start" this unit, and then remember to "enable" this unit to actually enable the corresponding reverse dependency.  For instance,
[Unit]
BindsTo=sys-subsystem-net-devices-%i.device
After=sys-subsystem-net-devices-%i.device
...
[Install]
WantedBy=sys-subsystem-net-devices-%i.device

---

Similarly, modify the "Requires=" description to disclose that that option is also not useful when referencing a "device" type unit.

---

Also note that "Requires=" is not useful with "device" type units because the dependency would probably  be the reverse of what you intended, where systemd cannot "start" a nonexistent device.  Instead, use "BindsTo=" in the Unit Section to "stop" this unit and "WantedBy=" in the Install Section to "start" this unit, and then remember to "enable" this unit to actually enable the corresponding reverse dependency.

---

Similarly, modify the "BindsTo=" description to disclose the method required to enable the reverse dependency.

---

Note that "BindsTo=" will only "stop" this unit, and will not "start" it.  To create a reverse dependency which can "start" this unit, use "WantedBy=" in the Install Section, and then remember to "enable" this unit to actually enable the corresponding reverse dependency.

---

**Comment by heftig** [comment]
Units can be generated from many sources - the DBus interface even lets you create transient `service` units that need no configuration files backing them because the needed configuration is  supplied with the call.

When units are loaded (either when first referenced or when systemd reloads) the configuration files for them are read and merged. This also applies to `device` units. Loading will fail if the unit must be configured to be valid (like `service` units) but no configuration was found.

"Starting" a `device` unit waits for the device to appear. "Stopping" a `device` unit waits for the device to disappear. These are subject to the usual job timeouts.

I guess it might be a good idea to clarify what causes jobs to be generated - other jobs (like for `PartOf`) or certain events (like for `BindsTo`). A unit might enter stopped state (for example, a `service` decides to terminate or a `device` is unplugged) without generating the jobs that a `systemctl stop` would.

Having to use `WantedBy` and `systemctl enable` to create a reverse activating dependency is a consequence of systemd lazily loading units. It cannot know whether there exists an unloaded unit that has such a dependency. `enable` creates a forward dependency that systemd can follow.

**Comment by thx1111** [comment]
Jan, can you say if my statement about IPC Mechanisms is complete, or over-broad?  Does systemd actually ever track device files in /dev/, or does systemd only ever have to track sysfs files in /sys/?  Similarly, does systemd ever actually track named pipes?  I assume that systemd never has reason to track an unnamed pipe.

Also, is there some systemctl command that will test for the consistency of dependency rules in the sum of all Unit Configuration Files, and also disclose dependency loops?  Right now, the only way I know to do this is to reboot the machine and then look for systemd messages about ordering cycles, event loops, and such.

**Comment by heftig** [comment]
systemd gets its device info from udev, which in turn gets notifications from the kernel.

Named pipes (FIFOs) can be created via `socket` units and passed as FDs to services just like sockets.

Perhaps try `/usr/lib/systemd/systemd --test --system --unit=multi-user.target`, which will build the initial transaction, dump it and exit.

**Comment by thx1111** [comment]
> Perhaps try /usr/lib/systemd/systemd --test --system --unit=multi-user.target

Hmm - that seems to be too much information.  There should be some more concise command to test for the consistency of dependency rules, noting only the warnings and errors, don't you think?

> systemd gets its device info from udev, which in turn gets notifications from the kernel.

Aha - I see now that "device" file names will begin both with "dev-" and with "sys-", and "device" Units can refer to special files in both /dev/ and /sys/.

**Comment by thx1111** [comment]
The explicit definition of "Unit" should also include "D-bus" in the exclusive list of IPC mechanisms:

3) An Inter-Process Communication Mechanism of a type limited to file, socket, named pipe, and D-bus type mechanisms.

**Comment by poettering** [comment]
I am sorry, but this is really confusing. We never use the term "unit" for anything but the unit types such as service, target, socket, mount, device, automount, busname, and so on. And that's defined in the systemd man page where all the types are listed with references documenting their specific details.

**Comment by thx1111** [comment]
Hmm - I also am noticing that the "After=" dependency option is meaningless in the context of a "device" type Unit.  Since a device interface is "created" by udev, not systemd, there is no "device" for systemd to "start" in sequence.

So then a note to that effect should be added to the "After=" description in systemd.unit(5).

**Comment by thx1111** [comment]
> I am sorry, but this is really confusing. We never use the term "unit" for anything but the unit types

In the man page, systemd.unit(5), we can see immediately, "A unit configuration file encodes information about..." and then shortly following, "In addition to the generic [Unit] and [Install] sections described here, each unit may have a type-specific section, ...".  And further on , "Units listed in this option..."  And in systemd(1), "The following unit types are available..." and later, "unit directories".  It seems apparent that the term "unit" is used to refer to several different kinds of things, as I have tried to enumerate.

In systemd(1), we see the statement "Service units, which start and control daemons..."  "Service units, which start and control daemons" are _not_ the same thing as the daemons that the "units" start and control.  Since that statement implies that a "unit" is not the "daemon" itself, I see that the definition that I proposed above is inconsistent with that statement, and I am back to asking "What _is_ a Unit?"

Certain characteristics can be inferred _about_ a "Unit".  It implicitly has a "name", which can be "listed".  It has a "configuration file" and a "section" and a "directory" and a "type".  But, what _is_ it?  A "has a" quality is not the same as an "is a" quality.

Also in systemd(1), we see the word "unit" is associated with the thing upon which the "unit" is acting, as in the verbs: starting, controlling, encapsulating, grouping, exposing, providing, saving, triggering, activating, and managing.  But "subject" type things are quite distinct from "object" type things.  The man page describes some "object" type things, but does not explicitly define the "subject" type thing which is acting upon the "object" type things.

From systemd(1): "Units encapsulate objects", "Units are configured", "But some units are created automatically, dynamically, or programmatically".  We are not told how the "other" units might be "created".  Nor are  we told precisely _what_ is being "created".

Perhaps you mean that a "Unit" is a set of heuristics used to effect an end result?  Perhaps a "Unit" is a particular subset of heuristics or a particular set of functions encoded in the program "systemd"?  Or in some other program?  Perhaps a "Unit" is an instance of a "state machine" internal to, and managed by, the program "systemd", which corresponds to a specific process or IPC mechanism?

Perhaps a "Unit" is an example of a "predictor" for a process or IPC mechanism, in the Model Predictive Control sense, from the field of Control Theory?

We can see also in systemd(1) the statement "Units are named as their configuration files."  This is what is most confusing, this "overloading" of names, "overloading" the set of things to which a particular name can refer.  Whenever the documentation refers to a "Unit", are we talking about some complex subset of heuristic functions?  Or are we talking about a Configuration File?  A system administrator will be familiar with a "Configuration File", but not so much with "an automatically created subset of heuristic functions."

More significantly, without properly knowing what a "Unit" _is_, there is a temptation to suppose that a "Unit" _is_ the process or the IPC mechanism _associated_ with the unit.  But, as the documentation requires, that supposition would be wrong.

We can search the Internet for "What are systemd units?"  There we will find vague allusions to "resources" and "objects", but that is not much more informative than saying that "Units" are "things".

Do this, run "systemctl --all --type=device list-units".  There are a list of names, states, and descriptions.  What _are_ those things???  They are not filenames.  They are not special files.  Or run "systemctl --all --type=service list-units".  Those are not processes, though they look kind of like processes.

Or just run "systemctl --all list-units".  Those things must represent "Units".  But they are not "processes", or "IPC mechanisms".  They are not any kind of traditional "Unix" thing.  It would be nice, some day, to formally articulate what these "Unit" things _are_.  They are something new, or at least, new to system administration.

**Comment by arvidjaar** [comment]
> It would be nice, some day, to formally articulate what these "Unit" things are.

"Unit" is a unit of systemd dependency/lifetime tracking. Nothing more. If you have suggestion how to put it in proper English, you are welcome.

**Comment by thx1111** [comment]
> "Unit" is a unit...

That is "circular logic".

> "Unit" is a ... tracking.

That may be the essence, though the word "tacking" does not convey the entire functionality of a Unit.  The task would then be to formally describe the nature of systemd "tracking".  And then, different Unit "types" have differing functionality.  So there is what a Unit is, in general, distinct from the differing functions of the various Unit "types".  This is an important issue because different type units have different types of state transitions, and these different types of state transitions are recognized selectively by the various dependency options.

In particular, the man page systemd.device(5) says explicitly "This unit type has no specific options.", which is misleading.  Device Units _do_ have dependency options that specifically have _no_ effect, as for instance, where the "PartOf=" option applies only to Units that "start" and "stop" and where Device Units do not "start" or "stop" and instead only "appear" and "disappear", in the systemd lingo.

> If you have suggestion how to put it in proper English, you are welcome.

Yes, I have offered a few different things so far, but there has not been much feedback.  I am still forming an understanding of how systemd is structured, conceptually.  Let me try again. Today, I will suggest that "systemd" is a Declarative Programming Language, and that a systemd Unit is a set of rules and facts expressed in that language and associated with a specific process or IPC mechanism, and then note that systemd automatically provides a default set of rules and facts for each process and IPC mechanism where each rule set can then be modified using Unit Configuration Files.  The "facts" represent the current state of a process or IPC mechanism, and the "rules" describe the dependencies between the various processes and IPC mechanisms.

Would you agree or disagree with that definition?

I will note that "file systems" and "partitions" should also be added to my exclusive list of IPC mechanisms, to account for mount points, auto mounts, and swap partitions,  Perhaps systemd "targets" should be considered a new and distinct type of IPC mechanism?  Or are they just an instance of a kind of special file in a unique file system?  I do not think that "runlevel" is included somewhere in /sys/ or tracked by the kernel.  I suppose that there is effectively the "systemdfs special file system" whose files can be listed using "systemctl --all list-units", and possibly including the special file "multi-user.target".

Altogether, systemd itself could be considered to be a Declarative Programming Language, a compiler for that language, and a special file system, as well as a Process manager, and an IPC mechanism manager.

**Comment by thx1111** [comment]
Along those lines, it may be more concise to say that a systemd Unit is a Program in the systemd Declarative Programming Language, and that systemd is the Interpreter for those programs.

**Comment by keszybz** [comment]
The documentation has been improved in various ways. In particular 2bf9250617d adds an explanation of reverse deps.
